### ISSUE TITLE
 Block user But i can't stop notification in objective c

### ISSUE THREAD

**Issue by sohilmultani** [QuickBlox/quickblox-ios-sdk#805]
i have create an application for chatting. i have use QuickBlox to chat with each other.

How can i stop Notification when user are Block

i have register two user and chat with each other it will work perfect. But when user A Block user B it will stop the message from sending B but notification are not stop.

**i have send notification like this :**

```
- (void)didPressSendButton:(UIButton *)button
           withMessageText:(NSString *)text
                  senderId:(NSUInteger)senderId
         senderDisplayName:(NSString *)senderDisplayName
                      date:(NSDate *)date {

    NSMutableDictionary *payload = [NSMutableDictionary dictionary];
    NSMutableDictionary *aps = [NSMutableDictionary dictionary];
    [aps setObject:@"default" forKey:QBMPushMessageSoundKey];
    [aps setObject:text forKey:QBMPushMessageAlertKey];
    [payload setObject:aps forKey:QBMPushMessageApsKey];

    QBMPushMessage *pushMessage = [[QBMPushMessage alloc] initWithPayload:payload];
    NSMutableArray *mutableOccupants = [self.dialog.occupantIDs mutableCopy];
    [mutableOccupants removeObject:@([self senderID])];
    NSNumber *opponentID = [mutableOccupants firstObject];


    [QBRequest sendPush:pushMessage toUsers:[NSString stringWithFormat:@"%@",opponentID] successBlock:^(QBResponse *response, QBMEvent *event) {
            // Successful response with event
        NSLog(@"Noti Response : %@",response);
    } errorBlock:^(QBError *error) {
            // Handle error
    }];

} 

``` 
**Block USer**  
```
-(void)viewDidLoad {

   [[QBChat instance] addDelegate:self];
   NSMutableArray *mutableOccupants = [self.dialog.occupantIDs mutableCopy];
    [mutableOccupants removeObject:@([self senderID])];
    blockOpponentID = [[mutableOccupants firstObject] integerValue];

    [[QBChat instance] retrievePrivacyListWithName:[NSString stringWithFormat:@"%ld",(long)blockOpponentID]];
    [[QBChat instance] retrievePrivacyListNames];

}

#pragma mark - Retrieve privacy lists names
- (void)chatDidReceivePrivacyListNames:(NSArray *)listNames
{
    NSLog(@"Peivecy List:%@",listNames);
}

- (void)chatDidNotReceivePrivacyListNamesDueToError:(id)error {
    NSLog(@"chat Did Not Receive Privacy List Names Due To Error");
}

#pragma mark Create a privacy list or edit existing list
- (void)chatDidSetPrivacyListWithName:(NSString *)name {
    NSLog(@"chat Did Set Privacy List With Name : %@",name);
    [[QBChat instance] setDefaultPrivacyListWithName:name];
    [[QBChat instance] setActivePrivacyListWithName:name];
}

- (void)chatDidNotSetPrivacyListWithName:(NSString *)name error:(id)error {
    NSLog(@"chat Did Not Set Privacy List With Name: %@ due to error:%@", name, error);
}

#pragma mark Activate a privacy list
- (void)chatDidSetDefaultPrivacyListWithName:(NSString *)name{
    [[QBChat instance] setDefaultPrivacyListWithName:name];

    NSLog(@"chat Did Set Default Privacy List With Name : %@",name);
}

- (void)chatDidNotSetDefaultPrivacyListWithName:(NSString *)name error:(id)error{
    NSLog(@"chat Did Not Set Default Privacy List With Name %@",name);
    NSLog(@"chat Did Not Set Default Privacy List With Name %@",error);
}


#pragma mark Retrieve a privacy list

- (void)chatDidReceivePrivacyList:(QBPrivacyList *)privacyList  {

    NSLog(@"Peivecy List:%@",[[[privacyList privacyItems] objectAtIndex:0] valueForKey:@"action"]);
    NSInteger action =[[[[privacyList privacyItems] objectAtIndex:0] valueForKey:@"action"] integerValue];
    if (action==0)
    {
        NSLog(@"Block");
        blockUnblockStatus=@"Block";
    }
    else
    {
        NSLog(@"Unblock");
        blockUnblockStatus=@"Unblock";
    }
}

- (void)chatDidNotReceivePrivacyListWithName:(NSString *)name error:(id)error {
    NSLog(@"Peivecy error:%@",error);
    blockUnblockStatus=@"Block";
}

#pragma mark - Delete existing privacy list

- (void)chatDidRemovedPrivacyListWithName:(NSString *)name {
    NSLog(@"chatDidRemovedPrivacyListWithName %@", name);
  }
```

**Block Unblock User Like this**
```
- (void)blockUserWithQBId:(NSUInteger)qbID
{
    QBPrivacyItem *item = [[QBPrivacyItem alloc] initWithType:USER_ID valueForType:qbID action:DENY];
    if (_blockPrivacyList)
    {
        [_blockPrivacyList addObject:item]; // add new user if already privacy list is there
    }else
        _blockPrivacyList = [[QBPrivacyList alloc] initWithName:[NSString stringWithFormat:@"%lu",(unsigned long)qbID] items:@[item]]; // create new privacy list if not before created
    [[QBChat instance] setPrivacyList:_blockPrivacyList];
    [[QBChat instance] setActivePrivacyListWithName:[NSString stringWithFormat:@"%lu",(unsigned long)qbID]];
    blockUnblockStatus=@"Unblock";
}

- (void)UnblockUserWithQBId:(NSUInteger)qbID
{
    QBPrivacyItem *item = [[QBPrivacyItem alloc] initWithType:USER_ID valueForType:qbID action:ALLOW];
    if (_blockPrivacyList)
    {
        [_blockPrivacyList addObject:item];

    }else
        _blockPrivacyList = [[QBPrivacyList alloc] initWithName:[NSString stringWithFormat:@"%lu",(unsigned long)qbID] items:@[item]]; // create new privacy list if not before created

    [[QBChat instance] setPrivacyList:_blockPrivacyList];
    [[QBChat instance] setActivePrivacyListWithName:[NSString stringWithFormat:@"%lu",(unsigned long)qbID]];
    blockUnblockStatus=@"Block";
}

```

**Comment by VitaliyGuru** [comment]
@sohilmultani, Privacy list blocks only chat alerts notifications. (Automatic alerts)
Manually sent notifications (e.g. via the request with QBMPushMessage) are not blocked.
Try to send notification in completion block of message sending.
```objective-c
QBChatDialog *dialog;
QBChatMessage *message;
[dialog sendMessage:message completionBlock:^(NSError * _Nullable error) {
    if (!error) {
        //send notification, blocked user will receive error
    }
}];
```

**Comment by VitaliyGuru** [comment]
@sohilmultani, Does chat alerts not satisfy your needs? Why do you send push with the message?

**Comment by Pro100Andrey** [comment]
Hi @sohilmultani, Any updates?

**Comment by sohilmultani** [comment]
Hello @VitaliyGuru and @Pro100Andrey 

its not work i have put this code in 

`QMChatService.m`

 ```
[dialog sendMessage:message completionBlock:^(NSError *error) {
        
        __typeof(weakSelf)strongSelf = weakSelf;
        
        if (error == nil) {
            
        
            NSMutableDictionary *payload = [NSMutableDictionary dictionary];
            NSMutableDictionary *aps = [NSMutableDictionary dictionary];
            [aps setObject:@"default" forKey:QBMPushMessageSoundKey];
            [aps setObject:message.text forKey:QBMPushMessageAlertKey];
            [payload setObject:aps forKey:QBMPushMessageApsKey];
            
            QBMPushMessage *pushMessage = [[QBMPushMessage alloc] initWithPayload:payload];
            
            
            NSString *ikid =[NSString stringWithFormat:@"%d",currentUser.ID];
            
            // Send push to users with ids
            NSMutableArray *mutableOccupants = [dialog.occupantIDs mutableCopy];
            [mutableOccupants removeObject:[NSString stringWithFormat:@"%d",currentUser.ID]];
            NSNumber *opponentID = [mutableOccupants firstObject];
            
            
            [QBRequest sendPush:pushMessage toUsers:[NSString stringWithFormat:@"%@",opponentID] successBlock:^(QBResponse *response, QBMEvent *event) {
                // Successful response with event
            } errorBlock:^(QBError *error) {
                // Handle error
            }];
}
```

**Comment by soulfly** [comment]
Fixed, so closing it

**Comment by ZalakSA** [comment]
@soulfly Hello,

I am also facing same issue. Can you please help me, how you made it working?

I am writing push notification code inside completion handler.

Thanks,
Zalak

**Comment by VitaliyGuru** [comment]
@ZalakSA, Please create a separate issue and describe it in details.
